File: src/main/java/io/github/liquidTruffle/lexer/Lexer.java
package io.github.liquidTruffle.lexer;

import java.io.Reader;
import java.io.StringReader;
import java.util.*;

public class Lexer implements TokenStream {
    private final Reader reader;
    private final char[] buffer = new char[4096]; // Buffer for reading from Reader
    private int bufferEnd = 0;
    private int bufferPos = 0;
    private int position = 0; // Global position counter
    private boolean eof = false;
    private LexerMode currentMode = LexerMode.IN_TEXT;
    
    // Streaming state
    private Token nextToken = null;
    private boolean hasNextToken = false;
    private boolean streamInitialized = false;
    
    // Liquid reserved keywords
    private final Set<String> keywords = new HashSet<>(Arrays.asList(
        // Control flow tags
        "assign", "capture", "case", "comment", "cycle", "for", "in", "break", "continue",
        "if", "include", "raw", "unless", "endfor", "endif", "endunless", "endcase",
        "else", "elsif", "when", "tablerow", "endtablerow", "increment", "decrement",
        "liquid", "echo", "render", "section", "endsection", "schema", "form",
        "paginate", "endpaginate", "layout", "block", "endblock", "extends",
        // Logical operators
        "and", "or", "not", "contains", "equals", "greater_than", "less_than",
        "greater_than_or_equal_to", "less_than_or_equal_to", "not_equals",
        // Other keywords
        "true", "false", "nil", "null", "empty", "blank", "default", "with",
        "first", "last", "size"
    ));

    public Lexer(Reader reader) {
        this.reader = reader;
    }
    
    public Lexer(String src) {
        this(new StringReader(src));
    }
    
    /**
     * This loads all tokens in memory as a list. Please do not use other than for tests or very simple lexing.
     */
    public List<Token> lex() {
        List<Token> tokens = new ArrayList<>();
        fillBuffer(); // Initial buffer fill

        while (!eof || bufferPos < bufferEnd) {
            Token token = processCurrentState();
            tokens.add(token);
            fillBuffer();
        }

        tokens.add(new Token(TokenType.EOF, "", position, position));
        return tokens;
    }

    // TokenStream interface implementation
    @Override
    public Token peek() {
        if (!streamInitialized) {
            initializeStream();
        }
        return nextToken;
    }
    
    @Override
    public Token advance() {
        if (!streamInitialized) {
            initializeStream();
        }
        Token current = nextToken;
        if (hasNextToken && current.type() != TokenType.EOF) {
            nextToken = getNextToken();
            hasNextToken = nextToken != null;
        } else {
            nextToken = null;
            hasNextToken = false;
        }
        return current;
    }
    
    @Override
    public boolean hasNext() {
        if (!streamInitialized) {
            initializeStream();
        }
        return hasNextToken;
    }
    
    @Override
    public Token[] lookAhead(int n) {
        // For simplicity, we'll implement a basic lookahead
        // In a more sophisticated implementation, we might want to cache tokens
        List<Token> tokens = new ArrayList<>();
        int originalPosition = position;
        int originalBufferPos = bufferPos;
        boolean originalEof = eof;
        LexerMode originalMode = currentMode;
        
        try {
            for (int i = 0; i < n; i++) {
                if (eof && bufferPos >= bufferEnd) {
                    break;
                }
                Token token = processCurrentState();
                if (token.type() == TokenType.EOF) {
                    break;
                }
                tokens.add(token);
                fillBuffer();
            }
            return tokens.toArray(new Token[0]);
        } finally {
            // Restore state
            position = originalPosition;
            bufferPos = originalBufferPos;
            eof = originalEof;
            currentMode = originalMode;
        }
    }
    
    private void initializeStream() {
        fillBuffer(); // Initial buffer fill
        nextToken = getNextToken();
        hasNextToken = (nextToken != null && nextToken.type() != TokenType.EOF);
        streamInitialized = true;
    }
    
    private Token getNextToken() {
        if (eof && bufferPos >= bufferEnd) {
            return new Token(TokenType.EOF, "", position, position);
        }
        Token token = processCurrentState();
        fillBuffer();
        return token;
    }

    private Token processCurrentState() {
        return switch (currentMode) {
            case IN_TEXT -> processTextMode();
            case IN_TAG -> processTagMode();
            case IN_OBJ -> processObjectMode();
        };
    }

    private Token processTextMode() {
        if (peek2("{{")) {
            return processObjOpen();
        } else if (peek2("{%")) {
            return processTagOpen();
        } else if (peek2("{#")) {
            // Process comment: {# comment #}
            processCommentOpen();
            processComment();
            processCommentClose();
            return processText(); // Return empty text token for comments
        } else {
            return processText();
        }
    }

    private Token processTagMode() {
        if (Character.isWhitespace(peekChar())) {
            skipWhitespace();
        }
        if (peek2("%}")) {
            return processTagClose();
        } else if (peekChar() == '"' || peekChar() == '\'') {
            return processString();
        } else if (Character.isDigit(peekChar()) || (peekChar() == '-' && Character.isDigit(peekChar(1)))) {
            return processNumber();
        } else if (peek2("..")) {
            return processRange();
        } else if (peekChar() == '.') {
            return processDot();
        } else if (peekChar() == '|') {
            return processPipe();
        } else if (peekChar() == ',') {
            return processComma();
        } else if (peekChar() == ':') {
            return processColon();
        } else if (peekChar() == '?') {
            return processQuestion();
        } else if (peek2(">=")) {
            return processGTE();
        } else if (peek2("<=")) {
            return processLTE();
        } else if (peek2("==")) {
            return processEQ();
        } else if (peek2("!=")) {
            return processNE();
        } else if (peek2("**")) {
            return processExponent();
        } else if (peek2("..")) {
            return processRange();
        } else if (peekChar() == '>') {
            return processGT();
        } else if (peekChar() == '<') {
            return processLT();
        } else if (peekChar() == '+') {
            return processPlus();
        } else if (peekChar() == '-') {
            return processMinus();
        } else if (peekChar() == '*') {
            return processMultiply();
        } else if (peekChar() == '/') {
            return processDivide();
        } else if (peekChar() == '%') {
            return processModulo();
        } else if (peekChar() == '(') {
            return processLParen();
        } else if (peekChar() == ')') {
            return processRParen();
        } else if (peekChar() == '[') {
            return processLBracket();
        } else if (peekChar() == ']') {
            return processRBracket();
        } else if (peekChar() == '{') {
            return processLBrace();
        } else if (peekChar() == '}') {
            return processRBrace();
        } else if (isIdentStart(peekChar())) {
            return processIdentifier();
        } else {
            throw new LexerException("Invalid character in tag mode: " + peekChar());
        }
    }

    private Token processObjectMode() {
        if (Character.isWhitespace(peekChar())) {
            skipWhitespace();
        }
        if (peek2("}}")) {
            return processObjClose();
        } else if (peekChar() == '"' || peekChar() == '\'') {
            return processString();
        } else if (Character.isDigit(peekChar()) || (peekChar() == '-' && Character.isDigit(peekChar(1)))) {
            return processNumber();
        } else if (peek2("..")) {
            return processRange();
        } else if (peekChar() == '.') {
            return processDot();
        } else if (peekChar() == '|') {
            return processPipe();
        } else if (peekChar() == ',') {
            return processComma();
        } else if (peekChar() == ':') {
            return processColon();
        } else if (peekChar() == '?') {
            return processQuestion();
        } else if (peek2(">=")) {
            return processGTE();
        } else if (peek2("<=")) {
            return processLTE();
        } else if (peek2("==")) {
            return processEQ();
        } else if (peek2("!=")) {
            return processNE();
        } else if (peek2("**")) {
            return processExponent();
        } else if (peek2("..")) {
            return processRange();
        } else if (peekChar() == '>') {
            return processGT();
        } else if (peekChar() == '<') {
            return processLT();
        } else if (peekChar() == '+') {
            return processPlus();
        } else if (peekChar() == '-') {
            return processMinus();
        } else if (peekChar() == '*') {
            return processMultiply();
        } else if (peekChar() == '/') {
            return processDivide();
        } else if (peekChar() == '%') {
            return processModulo();
        } else if (peekChar() == '(') {
            return processLParen();
        } else if (peekChar() == ')') {
            return processRParen();
        } else if (peekChar() == '[') {
            return processLBracket();
        } else if (peekChar() == ']') {
            return processRBracket();
        } else if (peekChar() == '{') {
            return processLBrace();
        } else if (peekChar() == '}') {
            return processRBrace();
        } else if (isIdentStart(peekChar())) {
            return processIdentifier();
        } else {
            throw new LexerException("Invalid character in object mode: " + peekChar());
        }
    }

    private boolean isIdentStart(char c) {
        return Character.isLetter(c) || c == '_' || c == '-';
    }
    
    private boolean isIdentPart(char c) {
        return Character.isLetterOrDigit(c) || c == '_' || c == '-';
    }
    
    private String collectIdent() {
        StringBuilder result = new StringBuilder();
        if (isIdentStart(peekChar())) {
            result.append(peekChar());
            advance(1);
        }
        result.append(collectWhile(this::isIdentPart));
        return result.toString();
    }
    
    private Token parseString() {
        int startPos = position;
        char quote = peekChar();
        advance(1); // consume opening quote
        
        StringBuilder result = new StringBuilder();
        while (!eof || bufferPos < bufferEnd) {
            char c = peekChar();
            if (c == quote) {
                advance(1); // consume closing quote
                break;
            }
            if (c == '\\') {
                advance(1); // consume backslash
                if (!eof || bufferPos < bufferEnd) {
                    char next = peekChar();
                    result.append(next);
                    advance(1);
                }
            } else {
                result.append(c);
                advance(1);
            }
        }
        
        return new Token(TokenType.STRING, result.toString(), startPos, position);
    }
    
    private String collectText() {
        StringBuilder result = new StringBuilder();
        while (!eof || bufferPos < bufferEnd) {
            char c = peekChar();
            if (c == '\u0000') break; // EOF
            
            // Stop at special sequences
            if (peek2("{{") || peek2("{%") || peek2("{#")) {
                break;
            }
            
            result.append(c);
            advance(1);
        }
        return result.toString();
    }
    
    private String collectComment() {
        StringBuilder result = new StringBuilder();
        while (!eof || bufferPos < bufferEnd) {
            char c = peekChar();
            if (c == '\u0000') break; // EOF
            
            // Stop at comment close
            if (peek2("#}")) {
                break;
            }
            
            result.append(c);
            advance(1);
        }
        return result.toString();
    }

    // Processing functions
    private Token processObjOpen() {
        int startPos = position;
        advance(2);
        
        // Check for whitespace control
        if (peekChar() == '-') {
            advance(1);
            currentMode = LexerMode.IN_OBJ;
            return new Token(TokenType.OBJECT_OPEN_WS, "{{-", startPos, position);
        }
        
        currentMode = LexerMode.IN_OBJ;
        return new Token(TokenType.OBJECT_OPEN, "{{", startPos, position);
    }

    private Token processObjClose() {
        int startPos = position;
        
        // Check for whitespace control before advancing
        if (peekChar(-1) == '-') {
            advance(2);
            currentMode = LexerMode.IN_TEXT;
            return new Token(TokenType.OBJECT_CLOSE_WS, "-}}", startPos - 1, position);
        }
        
        advance(2);
        currentMode = LexerMode.IN_TEXT;
        return new Token(TokenType.OBJECT_CLOSE, "}}", startPos, position);
    }

    private Token processTagOpen() {
        int startPos = position;
        advance(2);
        
        // Check for whitespace control
        if (peekChar() == '-') {
            advance(1);
            currentMode = LexerMode.IN_TAG;
            return new Token(TokenType.TAG_OPEN_WS, "{%-", startPos, position);
        }
        
        currentMode = LexerMode.IN_TAG;
        return new Token(TokenType.TAG_OPEN, "{%", startPos, position);
    }
    
    private Token processTagClose() {
        int startPos = position;
        
        // Check for whitespace control before advancing
        if (peekChar(-1) == '-') {
            advance(2);
            currentMode = LexerMode.IN_TEXT;
            return new Token(TokenType.TAG_CLOSE_WS, "-%}", startPos - 1, position);
        }
        
        advance(2);
        currentMode = LexerMode.IN_TEXT;
        return new Token(TokenType.TAG_CLOSE, "%}", startPos, position);
    }
    
    private void skipWhitespace() {
        collectWhile(Character::isWhitespace);
    }
    
    private Token processPipe() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.PIPE, "|", startPos, position);
    }
    
    private Token processColon() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.COLON, ":", startPos, position);
    }
    
    private Token processComma() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.COMMA, ",", startPos, position);
    }
    
    private Token processDot() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.DOT, ".", startPos, position);
    }

    private Token processString() {
        return parseString();
    }
    
    private Token processNumber() {
        int startPos = position;
        StringBuilder lexeme = new StringBuilder();
        
        // Handle negative numbers
        if (peekChar() == '-') {
            lexeme.append('-');
            advance(1);
        }
        
        // Collect digits
        lexeme.append(collectWhile(Character::isDigit));
        
        // Handle floating point numbers
        if (peekChar() == '.' && Character.isDigit(peekChar(1))) {
            lexeme.append('.');
            advance(1);
            lexeme.append(collectWhile(Character::isDigit));
            return new Token(TokenType.FLOAT, lexeme.toString(), startPos, position);
        }
        
        return new Token(TokenType.NUMBER, lexeme.toString(), startPos, position);
    }
    
    private Token processIdentifier() {
        int startPos = position;
        String lexeme = collectIdent();
        TokenType tokenType = keywords.contains(lexeme) ? TokenType.KEYWORD : TokenType.IDENT;
        return new Token(tokenType, lexeme, startPos, position);
    }
    
    private Token processText() {
        int startPos = position;
        String lexeme = collectText();
        return new Token(TokenType.TEXT, lexeme, startPos, position);
    }
    
    private Token processComment() {
        int startPos = position;
        String lexeme = collectComment();
        return new Token(TokenType.TEXT, lexeme, startPos, position);
    }

    private Token processGT() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.GT, ">", startPos, position);
    }

    private Token processLT() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.LT, "<", startPos, position);
    }

    private Token processGTE() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.GTE, ">=", startPos, position);
    }

    private Token processLTE() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.LTE, "<=", startPos, position);
    }

    private Token processEQ() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.EQ, "==", startPos, position);
    }

    private Token processNE() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.NE, "!=", startPos, position);
    }

    // New token processing methods for missing Liquid constructs
    
    private Token processCommentOpen() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.COMMENT_OPEN, "{#", startPos, position);
    }
    
    private Token processCommentClose() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.COMMENT_CLOSE, "#}", startPos, position);
    }
    
    private Token processPlus() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.PLUS, "+", startPos, position);
    }
    
    private Token processMinus() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.MINUS, "-", startPos, position);
    }
    
    private Token processMultiply() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.MULTIPLY, "*", startPos, position);
    }
    
    private Token processDivide() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.DIVIDE, "/", startPos, position);
    }
    
    private Token processModulo() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.MODULO, "%", startPos, position);
    }
    
    private Token processExponent() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.EXPONENT, "**", startPos, position);
    }
    
    private Token processLParen() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.LPAREN, "(", startPos, position);
    }
    
    private Token processRParen() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.RPAREN, ")", startPos, position);
    }
    
    private Token processLBracket() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.LBRACKET, "[", startPos, position);
    }
    
    private Token processRBracket() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.RBRACKET, "]", startPos, position);
    }
    
    private Token processLBrace() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.LBRACE, "{", startPos, position);
    }
    
    private Token processRBrace() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.RBRACE, "}", startPos, position);
    }
    
    private Token processQuestion() {
        int startPos = position;
        advance(1);
        return new Token(TokenType.QUESTION, "?", startPos, position);
    }
    
    private Token processRange() {
        int startPos = position;
        advance(2);
        return new Token(TokenType.RANGE, "..", startPos, position);
    }

    // Buffer management methods
    private void fillBuffer() {
        if (bufferPos >= bufferEnd && !eof) {
            bufferPos = 0;
            try {
                int charsRead = reader.read(buffer);
                if (charsRead == -1) {
                    eof = true;
                    bufferEnd = 0;
                } else {
                    bufferEnd = charsRead;
                }
            } catch (Exception e) {
                throw new LexerException("Error reading from input", e);
            }
        }
    }

    private char peekChar() {
        fillBuffer();
        return bufferPos < bufferEnd ? buffer[bufferPos] : '\u0000';
    }
    
    private char peekChar(int offset) {
        fillBuffer();
        int pos = bufferPos + offset;
        return pos < bufferEnd ? buffer[pos] : '\u0000';
    }

    private boolean peek2(String s) {
        fillBuffer();
        if (bufferPos + s.length() > bufferEnd) {
            // Need to check across buffer boundary
            return peekString(s.length()).equals(s);
        }
        char[] target = s.toCharArray();
        for (int i = 0; i < s.length(); i++) {
            if (buffer[bufferPos + i] != target[i]) {
                return false;
            }
        }
        return true;
    }

    private String peekString(int length) {
        fillBuffer();
        StringBuilder result = new StringBuilder();
        int remaining = length;
        int pos = bufferPos;

        while (remaining > 0) {
            if (pos >= bufferEnd) {
                // Need more data
                try {
                    int charsRead = reader.read(buffer, bufferEnd, buffer.length - bufferEnd);
                    if (charsRead <= 0) {
                        break; // EOF
                    }
                    bufferEnd += charsRead;
                } catch (Exception e) {
                    throw new LexerException("Error reading from input", e);
                }
            }

            int available = bufferEnd - pos;
            int toRead = Math.min(remaining, available);
            result.append(buffer, pos, toRead);
            pos += toRead;
            remaining -= toRead;
        }

        return result.toString();
    }

    private void advance(int count) {
        for (int i = 0; i < count; i++) {
            fillBuffer();
            if (bufferPos < bufferEnd) {
                bufferPos++;
                position++;
            }
        }
    }

    private String collectWhile(java.util.function.Predicate<Character> predicate) {
        StringBuilder result = new StringBuilder();
        while (!eof || bufferPos < bufferEnd) {
            char c = peekChar();
            if (!predicate.test(c)) break;
            result.append(c);
            advance(1);
        }
        return result.toString();
    }
}

File: src/main/kotlin/io/github/liquidTruffle/lexer/Lexer.kt
package io.github.liquidTruffle.lexer

import java.io.Reader
import java.io.StringReader

enum class LexerMode {
    IN_TEXT,
    IN_OBJ,
    IN_TAG
}

class Lexer(private val reader: Reader, private val reportWhitespaceTokens: Boolean = false) {
	private val buffer = CharArray(4096) // Buffer for reading from Reader
	private var bufferStart = 0
	private var bufferEnd = 0
	private var bufferPos = 0
	private var position = 0 // Global position counter
	private var eof = false
    private var currentMode = LexerMode.IN_TEXT
    
    // Liquid reserved keywords
    private val keywords = setOf(
        // Control flow tags
        "assign", "capture", "case", "comment", "cycle", "for", "in", "break", "continue",
        "if", "include", "raw", "unless", "endfor", "endif", "endunless", "endcase",
        "else", "elsif", "when", "tablerow", "endtablerow", "increment", "decrement",
        "liquid", "echo", "render", "section", "endsection", "schema", "form",
        "paginate", "endpaginate", "layout", "block", "endblock", "extends",
        // Logical operators
        "and", "or", "not", "contains", "equals", "greater_than", "less_than",
        "greater_than_or_equal_to", "less_than_or_equal_to", "not_equals",
        // Other keywords
        "true", "false", "nil", "null", "empty", "blank", "default", "with",
        "limit", "offset", "reversed", "sort", "where", "group_by", "order",
        "first", "last", "size", "join", "split", "strip", "strip_html",
        "strip_newlines", "newline_to_br", "escape", "url_encode", "url_decode",
        "base64_encode", "base64_decode", "hmac_sha1", "hmac_sha256", "md5",
        "sha1", "sha256", "date", "time", "now", "today", "yesterday", "tomorrow"
    )

    private val objectOpenMatcher = TokenMatcherAdapter({peek2("{{")}, ::processObjOpen)
    private val objectCloseMatcher = TokenMatcherAdapter({peek2("}}")}, ::processObjClose)
    private val tagOpenMatcher = TokenMatcherAdapter({peek2("{%")}, ::processTagOpen)
    private val tagCloseMatcher = TokenMatcherAdapter({peek2("%}")}, ::processTagClose)
    private val whitespaceMatcher = TokenMatcherAdapter({peek().isWhitespace()}, ::processWhitespace)
    private val pipeMatcher = TokenMatcherAdapter({peek() == '|'}, ::processPipe)
    private val colonMatcher = TokenMatcherAdapter({peek() == ':'}, ::processColon)
    private val dotMatcher = TokenMatcherAdapter({peek() == '.'}, ::processDot)
    private val commaMatcher = TokenMatcherAdapter({peek() == ','}, ::processComma)
    private val stringMatcher = TokenMatcherAdapter({peek() == '"' || peek() == '\''}, ::processString)
    private val numberMatcher = TokenMatcherAdapter({peek().isDigit()}, ::processNumber)
    private val identMatcher = TokenMatcherAdapter({isIdentStart(peek())}, ::processIdentifier)
    private val textMatcher = TokenMatcherAdapter({peek() != '\u0000'}, ::processText)
    private val gteMatcher = TokenMatcherAdapter({peek2(">=")}, ::processGTE)
    private val lteMatcher = TokenMatcherAdapter({peek2("<=")}, ::processLTE)
    private val eqMatcher = TokenMatcherAdapter({peek2("==")}, ::processEQ)
    private val neMatcher = TokenMatcherAdapter({peek2("!=")}, ::processNE)
    private val gtMatcher = TokenMatcherAdapter({peek() == '>'}, ::processGT)
    private val ltMatcher = TokenMatcherAdapter({peek() == '<'}, ::processLT)
    private val errorMatcher = TokenMatcherAdapter({true}, { throw LexerException("Invalid code" + peek()) })

    private val textModeMatchers = listOf(objectOpenMatcher, tagOpenMatcher, textMatcher)
    private val objectModeMatchers = listOf(objectCloseMatcher, whitespaceMatcher, stringMatcher, numberMatcher,
        dotMatcher, pipeMatcher, commaMatcher, colonMatcher, gteMatcher, lteMatcher, eqMatcher, neMatcher, gtMatcher,
        ltMatcher, identMatcher, errorMatcher)
    private val tagModeMatchers = listOf(tagCloseMatcher, whitespaceMatcher, stringMatcher, numberMatcher, dotMatcher,
        pipeMatcher, commaMatcher, colonMatcher, gteMatcher, lteMatcher, eqMatcher, neMatcher, gtMatcher, ltMatcher,
        identMatcher, errorMatcher)

	constructor(src: String, reportWhitespaceTokens: Boolean = false) : this(StringReader(src), reportWhitespaceTokens)
	
	fun lex(): List<Token> {
		val tokens = mutableListOf<Token>()
		fillBuffer() // Initial buffer fill

		while (!eof || bufferPos < bufferEnd) {
			val token = processCurrentState()
            if (reportWhitespaceTokens || token.type != TokenType.WHITESPACE) {
                tokens.add(token)
            }
            fillBuffer()
		}

		tokens.add(Token(TokenType.EOF, "", position, position))
		return tokens
	}
	
	private fun processCurrentState(): Token {
        return when(currentMode) {
            LexerMode.IN_TEXT -> textModeMatchers.first(TokenMatcher::match).run()
            LexerMode.IN_TAG -> tagModeMatchers.first(TokenMatcher::match).run()
            LexerMode.IN_OBJ -> objectModeMatchers.first(TokenMatcher::match).run()
        }
	}

	private fun isIdentStart(c: Char): Boolean = c.isLetter() || c == '_' || c == '-'
	private fun isIdentPart(c: Char): Boolean = c.isLetterOrDigit() || c == '_' || c == '-'
	
	private fun collectIdent(): String {
		val result = StringBuilder()
		if (isIdentStart(peek())) {
			result.append(peek())
			advance(1)
		}
		result.append(collectWhile { isIdentPart(it) })
		return result.toString()
	}
	
	private fun parseString(): Token {
		val startPos = position
		val quote = peek()
		advance(1) // consume opening quote
		
		val result = StringBuilder()
		while (!eof || bufferPos < bufferEnd) {
			val c = peek()
			if (c == quote) {
				advance(1) // consume closing quote
				break
			}
			if (c == '\\') {
				advance(1) // consume backslash
				if (!eof || bufferPos < bufferEnd) {
					val next = peek()
					result.append(next)
					advance(1)
				}
			} else {
				result.append(c)
				advance(1)
			}
		}
		
		return Token(TokenType.STRING, result.toString(), startPos, position)
	}
	
	private fun collectText(): String {
		val result = StringBuilder()
		while (!eof || bufferPos < bufferEnd) {
			val c = peek()
			if (c == '\u0000') break // EOF
			
			// Stop at special sequences
			if (peek2("{{") || peek2("}}") || peek2("{%") || peek2("%}")) {
				break
			}
			
			result.append(c)
			advance(1)
		}
		
		// Fallback for single character
		if (result.isEmpty() && (!eof || bufferPos < bufferEnd)) {
			result.append(peek())
			advance(1)
		}
		
		return result.toString()
	}

    // Processing functions
	private fun processObjOpen(): Token {
		val startPos = position
		advance(2)
		currentMode = LexerMode.IN_OBJ
		return Token(TokenType.VAR_OPEN, "{{", startPos, position)
	}

	private fun processObjClose(): Token {
		val startPos = position
		advance(2)
        currentMode = LexerMode.IN_TEXT
		return Token(TokenType.VAR_CLOSE, "}}", startPos, position)
	}

    private fun processTagOpen(): Token {
		val startPos = position
		advance(2)
        currentMode = LexerMode.IN_TAG
		return Token(TokenType.TAG_OPEN, "{%", startPos, position)
	}
	
	private fun processTagClose(): Token {
		val startPos = position
		advance(2)
        currentMode = LexerMode.IN_TEXT
		return Token(TokenType.TAG_CLOSE, "%}", startPos, position)
	}
	
	private fun processWhitespace(): Token {
		val startPos = position
		val lexeme = collectWhile { it.isWhitespace() }
		return Token(TokenType.WHITESPACE, lexeme, startPos, position)
	}
	
	private fun processPipe(): Token {
		val startPos = position
		advance(1)
		return Token(TokenType.PIPE, "|", startPos, position)
	}
	
	private fun processColon(): Token {
		val startPos = position
		advance(1)
		return Token(TokenType.COLON, ":", startPos, position)
	}
	
	private fun processComma(): Token {
		val startPos = position
		advance(1)
		return Token(TokenType.COMMA, ",", startPos, position)
	}
	
	private fun processDot(): Token {
		val startPos = position
		advance(1)
		return Token(TokenType.DOT, ".", startPos, position)
	}

	private fun processString(): Token {
		val token = parseString()
		return token
	}
	
	private fun processNumber(): Token {
		val startPos = position
		val lexeme = collectWhile { it.isDigit() }
		return Token(TokenType.NUMBER, lexeme, startPos, position)
	}
	
	private fun processIdentifier(): Token {
		val startPos = position
		val lexeme = collectIdent()
		val tokenType = if (keywords.contains(lexeme)) {
			TokenType.KEYWORD
		} else {
			TokenType.IDENT
		}
		return Token(tokenType, lexeme, startPos, position)
	}
	
	private fun processText(): Token {
		val startPos = position
		val lexeme = collectText()
		return Token(TokenType.TEXT, lexeme, startPos, position)
	}

	private fun processGT(): Token {
		val startPos = position
		advance(1)
		return Token(TokenType.GT, ">", startPos, position)
	}

	private fun processLT(): Token {
		val startPos = position
		advance(1)
		return Token(TokenType.LT, "<", startPos, position)
	}

	private fun processGTE(): Token {
		val startPos = position
		advance(2)
		return Token(TokenType.GTE, ">=", startPos, position)
	}

	private fun processLTE(): Token {
		val startPos = position
		advance(2)
		return Token(TokenType.LTE, "<=", startPos, position)
	}

	private fun processEQ(): Token {
		val startPos = position
		advance(2)
		return Token(TokenType.EQ, "==", startPos, position)
	}

	private fun processNE(): Token {
		val startPos = position
		advance(2)
		return Token(TokenType.NE, "!=", startPos, position)
	}

    // Buffer management methods
    private fun fillBuffer() {
        if (bufferPos >= bufferEnd && !eof) {
            bufferStart = position
            bufferPos = 0
            val charsRead = reader.read(buffer)
            if (charsRead == -1) {
                eof = true
                bufferEnd = 0
            } else {
                bufferEnd = charsRead
            }
        }
    }

    private fun peek(): Char {
        fillBuffer()
        return if (bufferPos < bufferEnd) buffer[bufferPos] else '\u0000'
    }

    private fun peek2(s: String): Boolean {
        fillBuffer()
        if (bufferPos + s.length > bufferEnd) {
            // Need to check across buffer boundary
            return peekString(s.length) == s
        }
        return bufferPos + s.length <= bufferEnd &&
                buffer.copyOfRange(bufferPos, bufferPos + s.length).contentEquals(s.toCharArray())
    }

    private fun peekString(length: Int): String {
        fillBuffer()
        val result = StringBuilder()
        var remaining = length
        var pos = bufferPos

        while (remaining > 0) {
            if (pos >= bufferEnd) {
                // Need more data
                val charsRead = reader.read(buffer, bufferEnd, buffer.size - bufferEnd)
                if (charsRead == -1) {
                    break // EOF
                }
                bufferEnd += charsRead
            }

            val available = bufferEnd - pos
            val toRead = minOf(remaining, available)
            result.append(buffer, pos, toRead)
            pos += toRead
            remaining -= toRead
        }

        return result.toString()
    }

    private fun advance(count: Int) {
        repeat(count) {
            fillBuffer()
            if (bufferPos < bufferEnd) {
                bufferPos++
                position++
            }
        }
    }

    private fun collectWhile(predicate: (Char) -> Boolean): String {
        val result = StringBuilder()
        while (!eof || bufferPos < bufferEnd) {
            val c = peek()
            if (!predicate(c)) break
            result.append(c)
            advance(1)
        }
        return result.toString()
    }

    private interface TokenMatcher {
        fun match(): Boolean
        fun run(): Token
    }

    private class TokenMatcherAdapter(val matcher:() -> Boolean, val runner: () -> Token): TokenMatcher {
        override fun match(): Boolean {
            return matcher()
        }
        override fun run(): Token {
            return runner()
        }
    }

}
